platform/linux-generic/arch/aarch64/odp_crypto_armv8.c

```diff
@@ -146,6 +184,215 @@ null_crypto_routine(odp_packet_t pkt ODP_UNUSED,
+
+	if (armv8_aes_gcm_set_counter(iv_ptr, iv_bit_length, &cs) != 0) {
+		ODP_DBG("ARM Crypto: Failure while setting nonce\n");
+		ret = -1;
+		goto err;
+	}
+
+	/* Copy AAD in a stack to make sure that the ARM crypto library can
+	 * read it in 16 byte chunks. */
+	uint8_t aad[ARM_CRYPTO_MAX_AAD_LENGTH];
+
+	memcpy(aad, param->aad_ptr, session->p.auth_aad_len);
+
+	uint32_t seg_len = 0;
+	uint8_t *data = odp_packet_offset(pkt, in_pos, &seg_len, NULL);
+
+	if (odp_unlikely(odp_packet_is_segmented(pkt)) ||
+	    odp_unlikely(odp_packet_tailroom(pkt) < OOB_READ_LEN)) {
+		/* Packet is segmented or it may not be safe to read beyond
+		 * the end of packet data. Copy the cipher range to a
+		 * contiguous buffer. */
+		odp_packet_copy_to_mem(pkt, in_pos, in_len, local.buffer);
+
+		data = local.buffer;
+	}
+
+	if (armv8_enc_aes_gcm_from_state(&cs,
+					 aad, aad_bit_length,
+					 data, plaintext_bit_length,
+					 local.buffer,
+					 local.digest) != 0) {
+		ODP_DBG("ARM Crypto: AES GCM Encoding failed\n");
+		ret = -1;
+		goto err;
+	}
+
+	odp_packet_copy_from_mem(pkt, param->cipher_range.offset,
+				 param->cipher_range.length, local.buffer);
+
+	odp_packet_copy_from_mem(pkt, param->hash_result_offset,
+				 session->p.auth_digest_len, local.digest);
+
+err:
+	return ret < 0 ? ODP_CRYPTO_ALG_ERR_DATA_SIZE :
+			 ODP_CRYPTO_ALG_ERR_NONE;
+}
+
+static
+odp_crypto_alg_err_t aes_gcm_decrypt(odp_packet_t pkt,
+				     const odp_crypto_packet_op_param_t *param,
+				     odp_crypto_generic_session_t *session)
+{
+	armv8_cipher_state_t cs = {
+		.counter = {
+			.d = {0, 0}
+		}
+	};
+	uint8_t *iv_ptr;
+	uint8_t tag[16];
+	uint64_t iv_bit_length;
+	uint64_t plaintext_bit_length = param->cipher_range.length * 8;
+	uint64_t aad_bit_length = session->p.auth_aad_len * 8;
+	uint32_t in_pos = param->cipher_range.offset;
+	uint32_t in_len = param->cipher_range.length;
+	int ret = 0;
+
+	/* Fail early if cipher_range is too large */
+	if (in_len > ARM_CRYPTO_MAX_DATA_LENGTH) {
+		ODP_DBG("ARM Crypto: Packet size too large for requested operation\n");
+		ret = -1;
+		goto err;
+	}
+
+#if ODP_DEPRECATED_API
+	if (param->cipher_iv_ptr)
+		iv_ptr = param->cipher_iv_ptr;
+	else if (session->p.cipher_iv.data)
+		iv_ptr = session->cipher.iv_data;
+	else
+		return ODP_CRYPTO_ALG_ERR_IV_INVALID;
+
+	iv_bit_length = session->p.cipher_iv.length * 8;
+#else
+	iv_ptr = param->cipher_iv_ptr;
+	iv_bit_length = session->p.cipher_iv_len * 8;
+	ODP_ASSERT(session->p.cipher_iv_len == 0 || iv_ptr != NULL);
+#endif
+
+	cs.constants = &session->cc;
+
+	if (armv8_aes_gcm_set_counter(iv_ptr, iv_bit_length, &cs) != 0) {
+		ODP_DBG("ARM Crypto: Failure while setting nonce\n");
+		ret = -1;
+		goto err;
+	}
+
+	/* Copy current hash bytes to a buffer and clear it */
+	odp_packet_copy_to_mem(pkt, param->hash_result_offset,
+			       session->p.auth_digest_len, tag);
+
+	if (odp_unlikely(session->p.hash_result_in_auth_range))
+		_odp_packet_set_data(pkt, param->hash_result_offset, 0,
```

Comment:
This clearing is not really needed anymore. Now the API explicitly says that auth_range is ignored with AES-GCM. The authentication tag cannot be within the encrypted range as the ciphertext is there. The bytes that are not ciphered but only authenticated are not in the packet but in a memory buffer pointer to by the AAD parameter and the API does not say that the AAD buffer may overlap packet data buffer at the hash result offset. Because of all this, the clearing does not really accomplish anything with AES-GCM (but with AES-GMAC and individual hash algorithms it would be needed).

Reply:
Okay, I will remove this condition. Thanks!